__COMPILER_HALT_OFFSET__ undefined const false detection

avant1 opened the pull request:
Hello.
PHP allows to put raw data ignored by PHP compiler in PHP files and then retreive raw data offset using `__COMPILER_HALT_OFFSET__ ` constant. If [__halt_compiler()](http://php.net/manual/en/function.halt-compiler.php) function is called in current file `__COMPILER_HALT_OFFSET__ ` will be defined, but for now PHPStan returns the following error:

```
Constant __COMPILER_HALT_OFFSET__ not found.
```

This PR adds a test to demonstrate the issue.

For now I am not familiar with PHPStan source code enough to fix it (and I am not even sure if it is easy to fix), but I'll try to look later.

ondrejmirtes commented:
Hi, thanks for the report. This means that PHP returns `false` for `defined('__COMPILER_HALT_OFFSET__')`  ¯\_(ツ)_/¯ I will have to fix it on the PHPStan side.

lookyman commented:
Man is this a corner case or what! I can't imagine ever using this, but thanks for showing me something I didn't know before. TIL :)

avant1 commented:
@ondrejmirtes Indeed, `__COMPILER_HALT_OFFSET__` is defined only when there is a `__halt_compiler` statement call *in the same file* (where we check it using `defined()`).

@lookyman I've updated analyzed test file to show use case for this PHP feature.

I've also added a simple patch that should fix this issue.
To check if `__COMPILER_HALT_OFFSET__` is available in file we should check if `__halt_compiler` was called in this file. However I found no way to do this without parsing the entire file again to get list of all nodes. And this is unacceptable for performance reasons.

Is there any way to get access to nodes of already parsed file using `PHPStan\Broker\Broker` or `PHPStan\Analyser\Scope`?

ondrejmirtes commented on src/Broker/Broker.php:
```diff
@@ -228,11 +232,27 @@ public function hasConstant(\PhpParser\Node\Name $nameNode, Scope $scope): bool
+	{
+		//todo file is already parsed, it should not be parsed again
+		//todo I've just found no other way to check if __halt_compiler() statement was called in file
+		$parser = new DirectParser((new ParserFactory())->create(ParserFactory::PREFER_PHP7), new NodeTraverser());
```
Just inject `Parser` instance into the Broker itself, don't instantiate it yourself. And it will inject cached implementation by default 😊

ondrejmirtes commented on tests/PHPStan/Rules/Constants/data/compiler-halt-offset-const.php:
```diff
@@ -0,0 +1,15 @@
+
+function greet(string $name): string
+{
+	$template = file_get_contents(__FILE__, false, null, __COMPILER_HALT_OFFSET__);
```
Also please add a test when the constant is unknown.

ondrejmirtes commented on src/Broker/Broker.php:
```diff
@@ -470,11 +478,26 @@ public function hasConstant(\PhpParser\Node\Name $nameNode, ?Scope $scope): bool
+	private function fileHasCompilerHaltStatementCalls(string $pathToFile): bool
+	{
+		$nodes = $this->parser->parseFile($pathToFile);
+		$haltCompilerCalls = array_filter($nodes, function (Node $node) {
```
1) This could be more efficient - stop after first occurence, don't go through all nodes.
2) How exactly the halt compiler call works? Can it be nested inside an if or some other construct or is it always in the top level?

avant1 replied:
> This could be more efficient - stop after first occurence, don't go through all nodes.
Indeed.

> How exactly the halt compiler call works? Can it be nested inside an if or some other construct or is it always in the top level?

No it cannot be nested. As everything after `__halt_compiler()` is ignored by compiler. So if you try use it inside an `if` with opening curly bracket it will lead to invalid syntax cause closing bracket will be missing for compiler.

Few examples: 
https://3v4l.org/rWSHt
https://3v4l.org/oi3F4

So, is there a way to analyze only top-level nodes?

ondrejmirtes replied:
Yes, this is how you did it 😊

avant1 replied:
Hah, I didn't realized it 😄

ondrejmirtes commented on src/Broker/Broker.php:
```diff
@@ -470,11 +478,26 @@ public function hasConstant(\PhpParser\Node\Name $nameNode, ?Scope $scope): bool
 	{
-		return $this->resolveName($nameNode, function (string $name): bool {
+		return $this->resolveName($nameNode, function (string $name) use ($scope): bool {
+			$isCompilerHaltOffset = $name === '__COMPILER_HALT_OFFSET__';
```
If this is false, you can immediately skip looking for the halt call.

ondrejmirtes commented:
Thanks! 🎉
